File: src/lib/api/audio.ts
import { command } from "./client";
import type {
  AudioCaptureState,
  AudioDeviceInfo,
} from "../../types";

export function listAudioDevices(): Promise<AudioDeviceInfo[]> {
  return command<AudioDeviceInfo[]>("list_audio_devices");
}

export function getDefaultAudioDevice(): Promise<AudioDeviceInfo | null> {
  return command<AudioDeviceInfo | null>("get_default_audio_device");
}

export function startAudioCapture(): Promise<AudioCaptureState> {
  return command<AudioCaptureState>("start_audio_capture");
}

export function stopAudioCapture(): Promise<void> {
  return command<void>("stop_audio_capture");
}

export function getAudioCaptureState(): Promise<AudioCaptureState> {
  return command<AudioCaptureState>("get_audio_capture_state");
}

File: src-tauri/src/commands/audio.rs
//! Audio commands: device discovery and capture lifecycle.

use tauri::{AppHandle, Emitter, State};

use crate::audio::AudioDeviceInfo;
use crate::errors::CommandResult;
use crate::events;
use crate::state::AppState;

#[tauri::command]
pub fn list_audio_devices(
    _app: AppHandle,
    _state: State<'_, AppState>,
) -> CommandResult<Vec<AudioDeviceInfo>> {
    crate::audio::device::list_input_devices()
}

#[tauri::command]
pub fn get_default_audio_device(
    _app: AppHandle,
    _state: State<'_, AppState>,
) -> CommandResult<Option<AudioDeviceInfo>> {
    crate::audio::device::default_input_device()
}

#[tauri::command]
pub fn start_audio_capture(
    app: AppHandle,
    state: State<'_, AppState>,
) -> CommandResult<crate::audio::AudioCaptureState> {
    let device_id = state
        .settings
        .lock()
        .map(|s| s.audio.input_device_id.clone())
        .unwrap_or(None);
    let capture = state.audio.start(device_id.as_deref())?;
    let _ = app.emit(events::AUDIO_STARTED, &capture);
    Ok(capture)
}

#[tauri::command]
pub fn stop_audio_capture(app: AppHandle, state: State<'_, AppState>) -> CommandResult<()> {
    state.audio.stop();
    let _ = app.emit(events::AUDIO_STOPPED, ());
    Ok(())
}

#[tauri::command]
pub fn get_audio_capture_state(
    _app: AppHandle,
    state: State<'_, AppState>,
) -> CommandResult<crate::audio::AudioCaptureState> {
    Ok(state.audio.state())
}
